### PR TITLE
NIFI-8962: Add overflow strategy to DistributeLoad

### PR DESCRIPTION
**Pull request by markobean**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:
Adds an 'overflow' strategy to the DistributeLoad processor. In this strategy, flowfiles are sent to the highest weighted and available relationship. Unlike 'next available', flowfiles are not distributed among all available relationships; they are sent to only one unless/until that relationship becomes unavailable. Then, it chooses the next highest weighted relationship which is available.

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [x] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DistributeLoad.java [hunk 1]
```diff
@@ -491,4 +508,35 @@ public boolean requiresAllDestinationsAvailable() {
...
+            final int numRelationships = relationshipList.size();
+
+            boolean foundFreeRelationship = false;
+            Relationship relationship = null;
+            // Getting set of available relationships only once. This may miss a relationship that recently became available, but
+            // overall is more efficient than re-calling for every relationship evaluation
+            Set<Relationship> availableRelationships = context.getAvailableRelationships();
+
+            int weightedIndex = 0;
+            while (!foundFreeRelationship) {
```
Comment (michael81877) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DistributeLoad.java:526]: Is it possible that no relationship may be available for a long amount of time, thus looping here consuming a thread? Should there be a timeout mechanism or do you think that's overkill?

**Review by michael81877 — COMMENTED** [review]
LGTM overall. A fairly simple, low-risk contribution.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DistributeLoad.java (hunk 1 above)
Comment (markobean) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DistributeLoad.java:526]: I don't think that's a valid concern - but I had to test it to be sure. The only time one of the defined relationships would not be available is if all relationships have reached their backpressure limit. And, when this is the case, the processor is not scheduled. Therefore, there will always be at least one available relationship in the set.

Comment (michael81877) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DistributeLoad.java:526]: Right on!

**Comment by michael81877** [comment]
This LGTM

**Review by michael81877 — APPROVED** [review]

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DistributeLoad.java [hunk 2]
```diff
@@ -68,19 +69,31 @@
...
+    public static final String OVERFLOW = "overflow";
+
+    public static final AllowableValue STRATEGY_ROUND_ROBIN = new AllowableValue(ROUND_ROBIN, ROUND_ROBIN,
+            "Relationship selection is evenly distributed in a round robin fashion; all relationships must be available.");
+    public static final AllowableValue STRATEGY_NEXT_AVAILABLE = new AllowableValue(NEXT_AVAILABLE, NEXT_AVAILABLE,
+            "Relationship selection is distributed across all available relationships in order of their weight; at least one relationship must be available.");
+    public static final AllowableValue STRATEGY_LOAD_DISTRIBUTION_SERVICE = new AllowableValue(LOAD_DISTRIBUTION_SERVICE, LOAD_DISTRIBUTION_SERVICE,
+            "Relationship selection is distributed by supplied LoadDistributionService Controller Service; at least one relationship must be available.");
+    public static final AllowableValue STRATEGY_OVERFLOW = new AllowableValue(OVERFLOW, OVERFLOW,
+            "Relationship selection is the highest weighted relationship available without further distribution; at least one relationship must be available.");
```
[the remark this answers is not in the document]
Comment (markobean) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DistributeLoad.java]: Returning to a long-overdue PR. Updated the relationship description

**Comment by markobean** [comment]
Despite this being a relatively old PR, I built `main` with the commits in this PR cherry-picked to ensure there are no conflicts.

**Comment by markap14** [comment]
Thanks for pushing the changes @markobean looks good to me. Did some testing and everything checked out as well. +1 will merge to main

**Pull request merged**